Charon/filetypes/CuraPackage.py

```
@@ -0,0 +1,124 @@
+        material_path_alias = "/materials"
+        self._ensureRelationExists(virtual_path=material_path_alias, relation_type="material", origin="/package.json")
+        self._writeToAlias(material_path_alias, package_filename, material_data)
+
+    ##  Gets a list of paths to quality files in the package.
+    def getQualities(self) -> List[str]:
+        return self.listPaths("/qualities")
+
+    ##  Add a new quality file.
+    #   \param: quality_data The data of the quality file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addQuality(self, quality_data: bytes, package_filename: str) -> None:
+        quality_path_alias = "/qualities"
+        self._ensureRelationExists(virtual_path=quality_path_alias, relation_type="quality", origin="/package.json")
+        self._writeToAlias(quality_path_alias, package_filename, quality_data)
+
+    ##  Gets a list of paths to machine definition files in the package.
+    def getMachines(self) -> List[str]:
+        return self.listPaths("/definitions")
+
+    ##  Add a new machine definition file.
+    #   \param machine_data The data of the machine definition file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addMachine(self, machine_data: bytes, package_filename: str) -> None:
+        machine_path_alias = "/definitions"
+        self._ensureRelationExists(virtual_path=machine_path_alias, relation_type="definition", origin="/package.json")
+        self._writeToAlias(machine_path_alias, package_filename, machine_data)
+
+    ##  Gets a list of paths to plugins in the package.
+    def getPlugins(self) -> List[str]:
+        return self.listPaths("/plugins")
+
+    ##  Add a new plugin.
+    #   \param plugin_data The ZIP file containing the plugin files converted to bytes.
+    #   \param plugin_id The ID of the plugin within the package.
```

What happens if I add a plugin with the same ID twice? Does it override it? Does it throw an exception?

---

Good one. I think we should throw an exception as it's usually unintended.